charts/velero/templates/backupstoragelocation.yaml

```diff
@@ -1,4 +1,43 @@
 {{- if .Values.backupsEnabled }}
+{{- if and .Values.configuration.backupStorageLocation (kindIs "slice" .Values.configuration.backupStorageLocation) }}
+{{- range .Values.configuration.backupStorageLocation }}
+---
+apiVersion: velero.io/v1
+kind: BackupStorageLocation
+metadata:
+  name: {{ .name }}
+  namespace: {{ $.Release.Namespace }}
+  annotations:
+    "helm.sh/hook": post-install,post-upgrade,post-rollback
+    "helm.sh/hook-delete-policy": before-hook-creation
+  labels:
+    app.kubernetes.io/name: {{ include "velero.name" $ }}
+    app.kubernetes.io/instance: {{ $.Release.Name }}
+    app.kubernetes.io/managed-by: {{ $.Release.Service }}
+    helm.sh/chart: {{ include "velero.chart" $ }}
+spec:
+  credential: 
+    name: {{ .credential.name }}
+    key: {{ .credential.key }}
+  provider: {{ .provider }}
```

Comment:
I believe the changes I have made to the two templates and helpers.tpl retain the `.provider` behavior. I did some quick tests with `helm template` to attempt to confirm this. Hopefully a breaking change won't be required!

Reply:
@bsteinm2 I did a test with `helm template`. The accessMode does not render. Could you help check it?

```shell
helm template velero     --namespace=velero     --create-namespace     --set-file credentials.secretContents.cloud=credentials-velero     --set configuration.provider=aws     --set configuration.backupStorageLocation.name=default     --set configuration.backupStorageLocation.bucket=velero     --set configuration.backupStorageLocation.config.region=minio-default     --set configuration.backupStorageLocation.config.s3ForcePathStyle=true     --set configuration.backupStorageLocation.config.s3Url=http://minio-default.velero.svc.cluster.local:9000     --set configuration.backupStorageLocation.config.publicUrl=http://localhost:9000     --set snapshotsEnabled=true     --set deployNodeAgent=true     --set configuration.volumeSnapshotLocation.name=default     --set configuration.volumeSnapshotLocation.config.region=minio-default     --set initContainers[0].name=velero-plugin-for-aws     --set initContainers[0].image=velero/velero-plugin-for-aws:v1.5.3     --set initContainers[0].volumeMounts[0].mountPath=/target     --set initContainers[0].volumeMounts[0].name=plugins     charts/velero | vim -
```

Reply:
That's interesting.. I see accessMode rendering when I use any valid sample values file.

I haven't been able to recreate this test using your command yet, as when I run it, I get the error: `Error: non-absolute URLs should be in form of repo_name/path_to_chart, got: 
zsh: no matches found: initContainers[0].name=velero-plugin-for-aws`

Maybe it's due to different helm cli versions? I am on v3.10.1

Reply:
@bsteinm2 It's [zsh issue](https://github.com/vmware-tanzu/helm-charts/blob/08bf56b0d8864014c1a7b704e3a3d35bc10848a0/charts/velero/README.md#:~:text=Users%20of%20zsh%20might%20need%20to%20put%20quotes%20around%20key/value%20pairs.), not helm. Could you use bash shell instead and rerun the above command?

Reply:
I have replicated the issue now, thanks for the help. I'm currently confused as to how `accessMode: ReadWrite` gets rendered with current main when running this command, though...

There is no default value for accessMode in `helpers.tpl`... it seems it could only receive the `ReadWrite` value from `values.yaml`, but I am not aware of this behavior existing for a `helm template` command without the `--values` flag.

Not sure if it helps, but the value renders correctly when I ran your command with the addition of `--set configuration.backupStorageLocation.accessMode=ReadWrite`.

Reply:
My confusion has been cleared up after some reading into how helm digests values files. The issue appears to be that it tries to find `accessMode` in `values.yaml` and cannot, because this file has the BSL config in a slice data structure now. So, `accessMode` renders correctly when the values.yaml of currently main is in the local repo when running your command. Hope this explanation makes sense.

Currently trying to find a suitable solution to this, and I am open to suggestions. One I am considering is to add a check after line 26 to see if accessMode has a value, and if not, just supply "ReadWrite". Not sure if this would be good practice.

Reply:
I vote for adding a default value to line 26.

Reply:
I think my last commit has solved the issue.. I did change your command a bit so that the format of the configuration.backupStorageLocation matched that of the current default values file on this fork. Your command should work now with the old values file format (configuration.backupStorageLocation is a map, not a slice). It doesn't work with your command paired with the new values file, since it looks for an accessMode in the wrong spot, the command just conflicts with the new values file in this context. Your command does work with the new templates and the old values file. For backwards compatibility's sake, we could revert the values file format such that configuration.BSL and .VSL are maps and just add some comments saying you can make it a slice to use this new feature of multiple BSLs and VSLs. 

For context, here is the command I used that works with the new values file format. Again, your command works with the old values file, though.

```
helm template velero    --namespace=velero   --create-namespace   --set-file credentials.secretContents.cloud=credentials-velero   --set configuration.provider=aws   --set configuration.backupStorageLocation[0].name=default   --set configuration.backupStorageLocation[0].bucket=velero   --set configuration.backupStorageLocation[0].credential.name=blah   --set configuration.backupStorageLocation[0].credential.key=blah   --set configuration.backupStorageLocation[0].config.region=minio-default   --set configuration.backupStorageLocation[0].config.s3ForcePathStyle=true   --set configuration.backupStorageLocation[0].config.s3Url=http://minio-default.velero.svc.cluster.local:9000   --set configuration.backupStorageLocation[0].config.publicUrl=http://localhost:9000   --set snapshotsEnabled=true   --set deployNodeAgent=true   --set configuration.volumeSnapshotLocation[0].name=default   --set configuration.volumeSnapshotLocation[0].config.region=minio-default   --set initContainers[0].name=velero-plugin-for-aws   --set initContainers[0].image=velero/velero-plugin-for-aws:v1.5.3   --set initContainers[0].volumeMounts[0].mountPath=/target   --set initContainers[0].volumeMounts[0].name=plugins   . --debug
```

Reply:
I did a test with old command, it doesn't work.

I am checking is there anyway to convert slice to map so we can address the backward compatible issues. Do you know how to achieve it? @bsteinm2 

If we can't, I also consider we could bump the major version to deprecate old command.

Reply:
Nothing immediately comes to mind about converting from slice to map, but I'll give it some thought.

Can you try the old command again with this updated values file? If this doesn’t work I'm leaning towards agreeing that we just bump the major version.

Reply:
@bsteinm2 Here you are.

```console
helm template velero    --namespace=velero   --create-namespace   --set-file credentials.secretContents.cloud=credentials-velero   --set configuration.provider=aws   --set configuration.backupStorageLocation.name=default   --set configuration.backupStorageLocation.bucket=velero   --set configuration.backupStorageLocation.credential.name=blah   --set configuration.backupStorageLocation.credential.key=blah   --set configuration.backupStorageLocation.config.region=minio-default   --set configuration.backupStorageLocation.config.s3ForcePathStyle=true   --set configuration.backupStorageLocation.config.s3Url=http://minio-default.velero.svc.cluster.local:9000   --set configuration.backupStorageLocation.config.publicUrl=http://localhost:9000   --set snapshotsEnabled=true   --set deployNodeAgent=true   --set configuration.volumeSnapshotLocation.name=default   --set configuration.volumeSnapshotLocation.config.region=minio-default   --set initContainers[0].name=velero-plugin-for-aws   --set initContainers[0].image=velero/velero-plugin-for-aws:v1.5.3   --set initContainers[0].volumeMounts[0].mountPath=/target   --set initContainers[0].volumeMounts[0].name=plugins   charts/velero
```

Reply:
@jenting thanks! Okay, your command is now rendering `accessMode` for me with the values reverted to the map format for BSL and VSL. This could be a good way to ship this feature, and I could add some more comments to clarify the enhancement being shipped here. Does that work?

Reply:
@bsteinm2 I am a bit confused now. With the values reverted to the map format for BSL and VSL, how can the user configure multiple BSLs and VSLs?

Reply:
With the values file in the map format, we can preserve backwards compatibility and recommend that users add a `-` at the start of each entry under BSL and VSL to use this feature. I would add this detail to the values file comments.

I'm not sure this is the best route/follows convention of how to introduce a new feature for a helm chart... If you think it would be better, I am fine with bumping the major chart version and having values.yaml use the slice format for BSL and VSL. Thanks for your continued input.